From e4586850e992baf7499f55ec37fb5e6fdcc07f8b Mon Sep 17 00:00:00 2001
From: Stuart Menefy <stuart.menefy@codasip.com>
Date: Thu, 18 Jan 2024 14:01:08 +0000
Subject: [PATCH 07/17] hw/riscv/hobgoblin: Wire up an SD card to the SPI
 controller

Add the infrastructure to allow an SD card to be connected to the
SPI controller.

Signed-off-by: Stuart Menefy <stuart.menefy@codasip.com>
---
 hw/riscv/hobgoblin.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/hw/riscv/hobgoblin.c b/hw/riscv/hobgoblin.c
index ef4f8ea2be..3f0b253641 100644
--- a/hw/riscv/hobgoblin.c
+++ b/hw/riscv/hobgoblin.c
@@ -30,6 +30,9 @@
 #include "hw/boards.h"
 #include "hw/loader.h"
 #include "hw/sysbus.h"
+#include "hw/misc/unimp.h"
+#include "hw/sd/sd.h"
+#include "hw/ssi/ssi.h"
 #include "target/riscv/cpu.h"
 #include "hw/riscv/riscv_hart.h"
 #include "hw/intc/riscv_aclint.h"
@@ -230,6 +233,11 @@ static void hobgoblin_board_init(MachineState *machine)
     int num_entries;
     const int smp_cpus = 1;
     SysBusDevice *ss;
+    DeviceState *sd_dev;
+    SSIBus *bus;
+    DriveInfo *dinfo;
+    BlockBackend *blk;
+    DeviceState *card_dev;
 
     /* Initialize SOC */
     object_initialize_child(OBJECT(machine), "soc", &s->soc,
@@ -344,6 +352,20 @@ static void hobgoblin_board_init(MachineState *machine)
     }
     rom_add_blob_fixed_as("mrom.reset", reset_vec, sizeof(reset_vec),
                           memmap[HOBGOBLIN_MROM].base, &address_space_memory);
+
+    /* Connect an SD card to SPI */
+    ss = SYS_BUS_DEVICE(s->spi);
+    bus = (SSIBus *)qdev_get_child_bus(s->spi, "spi");
+    sd_dev = ssi_create_peripheral(bus, "ssi-sd");
+
+    dinfo = drive_get(IF_SD, 0, 0);
+    blk = dinfo ? blk_by_legacy_dinfo(dinfo) : NULL;
+    card_dev = qdev_new(TYPE_SD_CARD);
+    qdev_prop_set_drive_err(card_dev, "drive", blk, &error_fatal);
+    qdev_prop_set_bit(card_dev, "spi", true);
+    qdev_realize_and_unref(card_dev,
+                           qdev_get_child_bus(sd_dev, "sd-bus"),
+                           &error_fatal);
 }
 
 static void hobgoblin_machine_init(MachineClass *mc)
-- 
2.39.3

